### PR TITLE
Add the means to extract the contextual properties from HttpChannel, TcpChannel and TrasportChannel without excessive typecasting

### PR DESCRIPTION
**Pull request by reta**
### Description
While instrumenting the transports (both TCP and HTTP), we found out that `security` plugin heavily depends on `netty` transport (in fact, it does not support anything else) but it is excessive typecasting the different transport object to their `Netty4Xxx` equivalents for accessing handlers and attributes. It significantly complicates any instrumentation efforts and needs general mechanism to access contextual information without typecasting.

### Related Issues
Related to https://github.com/opensearch-project/security/issues/3515

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [X] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
@peternied @Gaganjuneja fyi

**Comment by cwperks** [comment]
Thank you @reta! There are a couple of areas that the security plugin is coupled to netty:

1. The security plugin's implementation of SSL relies on adding a step to the netty pipeline. The SSL Handler is the first step of the pipeline: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/ssl/transport/SecuritySSLNettyTransport.java#L263-L285
2. As part of https://github.com/opensearch-project/security/pull/3418, a step in the netty pipeline was added before decompression to verify the request based on headers and uses netty channel attributes

**Comment by reta** [comment]
> As far as I can see, there are no automated tests to verify the behavior amongst other http engines

Thanks @cwperks , correct, I intend to pick this one up in context of https://github.com/opensearch-project/OpenSearch/pull/9672 since the new transport will be facing all these issues right away. I think for the instrumentation part with  change we should be set, what do you think?

**Comment by cwperks** [comment]
> I think for the instrumentation part with this change we should be set, what do you think?

The security plugin would need to set attributes as well. Can that be added?

**Comment by reta** [comment]
> The security plugin would need to set attributes as well. Can that be added?

Discussed with @cwperks offline, we'll looking into that separately

**Review by peternied — COMMENTED** [review]
Thanks for making access to this more generic - I much prefer this approach rather than downstream 'peaking' of the types.  Couple of minor comments around the new interface.

File: modules/transport-netty4/src/main/java/org/opensearch/http/netty4/Netty4HttpChannel.java [hunk 1]
```diff
@@ -98,6 +98,22 @@ public Channel getNettyChannel() {
...
+        if (handler == null && inboundPipeline() != null) {
+            handler = inboundPipeline().get(name);
+        }
+
+        if (handler != null && clazz.isInstance(handler) == true) {
+            return (T) handler;
+        }
+
+        return null;
+    }
```
[the remark this answers is not in the document]
Comment (reta) [reply modules/transport-netty4/src/main/java/org/opensearch/http/netty4/Netty4HttpChannel.java:115]: The problem of `Optional.ofNullable` is that is not supposed to be used in this way: it creates temp objects for a sake of using method chaining. In contrast, when `Optional` is a return value, it works perfectly.

**Comment by reta** [comment]
@cwperks @peternied @Gaganjuneja the attributes issue is more complicated, looking if we even need that, but in any case the solution would be different and it would make sense to have a dedicated pull request for it.

**Comment by Gaganjuneja** [comment]
LGTM.

**Review by peternied — APPROVED** [review]

**Review by cwperks — APPROVED** [review]

**Review by scrawfor99 — APPROVED** [review]

**Pull request merged**